harness/tests/experiment/utils.py

```diff
@@ -266,6 +274,65 @@ def make_trial_controller_from_trial_implementation(
         workloads=workloads,
     )
 
+def run_pytorch_trial_controller_from_trial_impl(
+    trial_class: Type[det.pytorch.PyTorchTrial],
+    hparams: Dict,
+    scheduling_unit: int = 1,
+    trial_seed: int = random.randint(0, 1 << 31),
+    exp_config: Optional[Dict] = None,
+    checkpoint_dir: Optional[str] = None,
+    latest_checkpoint: Optional[str] = None,
+    steps_completed: int = 0,
+    expose_gpus: bool = False,
+    max_batches: int = 100,
+    min_checkpoint_batches: int = 0,
+    min_validation_batches: int = 0,
+) -> PyTorchTrialController:
+
+    assert issubclass(trial_class, PyTorchTrial), "pytorch test method called for non-pytorch trial"
+
+    if not exp_config:
+        assert hasattr(
+            trial_class, "_searcher_metric"
+        ), "Trial classes for unit tests should be annotated with a _searcher_metric attribute"
+        searcher_metric = trial_class._searcher_metric  # type: ignore
+        exp_config = make_default_exp_config(
+            hparams, scheduling_unit, searcher_metric, checkpoint_dir=checkpoint_dir
+        )
+    env = make_default_env_context(
```

Comment:
If the harness doesn't use an EnvContext with pytorch trial anymore (which imo it shouldn't) then the tests must not use EnvContext with pytorch trial.  Otherwise you are basically just testing a special codepath that only exists for tests, rather than testing the actual correct codepath.

And... I understand how much work it is what I'm suggesting you do.  And I'm sorry.  

If I were you I would not rewrite the tests accordingly until the very end.  Maybe rewrite a couple so you can have some automated tests of the new trainer api while you develop, but I'd probably save test-writing for the very last step.